marlowe-playground-client/src/Examples/JS/Contracts.purs

```diff
@@ -199,3 +199,149 @@ makeDeposit(alice, 10,
+    return amount
+}
+
+function minus(v1: Value, v2: Value): Value {
+    return SubValue(v1, v2)
+}
+
+function sendPayment(fromParty: Party, toCounterParty: Payee, tkn: Token, payoff: Value): (_: Contract) => Contract {
+    return cont => Pay(fromParty, toCounterParty, tkn, payoff, cont)
+}
+
+function Do(actions: Array<(_: Contract) => Contract>, last: Contract): Contract {
+    const foldRight = <A, B>(xs: Array<A>, zero: B) => (f: (b: B, a: A) => B): B => {
+        const len = xs.length;
+        if (len === 0) return zero;
+        else {
+            const last = xs[len - 1];
+            const inits = xs.slice(0, len - 1);
+            return foldRight(inits, f(zero, last))(f);
+        }
+    }
+    return foldRight(actions, last)((b, a) => a(b))
+}
+
+const contract = (() => {
+    const partyCollateralToken = Token("", "")
+    const partyCollateralAmount = value(1000)
+    const counterPartyCollateralToken = Token("", "")
+    const counterPartyCollateralAmount = value(1000)
+    const endDate = 1000
+    const partyCollateralDeposit = Deposit(party, party, partyCollateralToken, partyCollateralAmount)
+
+    const counterPartyCollateralDeposit = Deposit(counterParty, counterParty, counterPartyCollateralToken, counterPartyCollateralAmount)
+
+
+    const minValue = (val1: Value, val2: Value) => Cond(ValueGE(val1, val2), val2, val1)
+
+    return Do([
+        waitForEvent(partyCollateralDeposit, before(100), orElse(end)),
+        waitForEvent(counterPartyCollateralDeposit, before(100), orElse(end)),
+        waitForEvent(receiveValue("price1"), before(100), orElse(end)),
+        waitFor(endDate),
+        waitForEvent(receiveValue("price2"), before(endDate + 100), orElse(end)),
+        letValue("delta", minus(readValue("price1"), readValue("price2")))
+    ],  checkIf(ValueEQ(useValue("delta"), value(0)), 
+            thenDo(end),
+            elseDo(
+                checkIf(ValueLT(useValue("delta"), value(0)),
+                    thenDo(
+                        Do([
+                            letValue("absdelta", minus(value(0), useValue("delta"))),
+                            (() => {
+                                const payoff = minValue(useValue("absdelta"), partyCollateralAmount)
+                                return sendPayment(fromParty(party), toParty(counterParty), withToken(partyCollateralToken), amountOf(payoff))
+                            })()
+                        ], end)
+                    ),
+                    elseDo(
+                        Do([
+                            (() => {
+                                const payoff = minValue(useValue("delta"), partyCollateralAmount)
+                                return sendPayment(fromParty(party), toParty(counterParty), withToken(partyCollateralToken), amountOf(payoff))
+                            })()
+                        ], end)
+                    )
+                )
+            )
+        )
+    )
+})()
```

Comment:
You could inline this function, I see why you would want to do this if you had several contracts in the same file, but in this case...

Reply:
this one is actually inlined - `const contract = {...}` doesn't work in TS. So I would have to `function genContract() = {...}; const contract = genContract()`

Reply:
I mean you could move the const declarations outside, and instead of `return blah` have `const contract = blah`

Reply:
Oh - I kinda don't like that because at the moment declarations that are outside - are basically helpers, that I'm planning to move to higher-level language (let's call it MarloweJSX) - while things inside contract lambda are related to the logic of the contract. 

So it is my way to highlight that stylistically